### DISCUSSION TITLE
One step set-up or update of Freeplane add-on development environment

### DISCUSSION THREAD

**Discussion by HaaiHenkie** [freeplane/freeplane#258] [Ideas]
As an add-on developer I would like to develop against the latest stable version of Freeplane. Setting up or updating a Freeplane add-on development environment involves downloading de source code of the stable release, followed by several steps in order to get it imported into your IDE (e.g. Eclipse or IDEA IntelliJ). I am looking for ways to automate those steps, so that a first set-up is easier for beginners and an update to a new stable version becomes easy and fast.

Last year I started using the [Gradle Plugin for Freeplane Script and Add-on Development](https://github.com/freeplane/gradle-plugin) and turned my add-on project into a Gradle project. I found that I could import Freeplane and my add-on project as a composite gradle build, as you can read in [this thread](https://sourceforge.net/p/freeplane/discussion/758437/thread/ccb0ecb6e5/). This is progress because it allows me to set-up my add-on development environment with a single import. But it still involves downloading the source code of the stable Freeplane release and several other manual steps.

I was wondering if the following could work. Stable Freeplane versions are made available in a Maven repository. You could then add it as an dependency in your ``build.gradle`` file, leading to an automatic download of Freeplane source and binaries, making it available in the IDE. This would only make sense if you still could run Freeplane from your IDE for testing and debugging the add-on. I seriously doubt whether this is possible, but what is the harm of asking?

Best whishes,

Henk

**Comment by EdoFro** [comment]
Totally agree.

I haven't been able to develop any add-on using an IDE. I installed IntelliJ Idea to do this but I got lost in all the steps and messages.

I'm still using the wikdshell add-on (groovy console) and notepad++ to do all my coding.

**Reply by EdoFro** [reply]
Maybe @macmarrum can help us

please??? ;)

**Comment by macmarrum** [comment]
I use git to get Freeplane source code.
I use IntelliJ IDEA to create my scripts.
I still need to review https://github.com/freeplane/gradle-plugin and to create an add-on.

So I'm not sure how much I can help you, but I'm willing to try 😀

**Reply by EdoFro** [reply]
Thanks!

I will ask you for help the next time I try to set up the IDE. Not in the next days, but hopefully in the near future.

cheers,
edo

**Comment by HaaiHenkie** [comment]
I mainly use Eclipse and experimented with IntelliJ IDEA. I cannot remember I had go through a lot of steps and messages.

You would use git to clone Freeplane if you want to contribute to the development of Freeplane. I think using it for script or add-on development is not recommendable, because you are developing against code which is being developed and still needs to be tested. It would be possible to get the stable code by checking out the tag for the latest release. In my set-up, a gradle composite build, I would get the Freeplane git repository inside the tree of my own git repository. I do not think that is a smart thing to do. I also think this is not a suitable approach for beginners.

A very interesting discussion, a bit off-topic though. My question is: could making Freeplane available in a Maven repository be a solution to a one-step setup of a add-on development environment?

**Reply by EdoFro** [reply]
Sorry. I have no idea.

**Comment by dpolivaev** [comment]
It would be nice to have Freeplane available in a maven repo and runnable easily. However, there is an obstacle related to its origins namely to FreeMind.

FreeMind has used so-called plug-ins which allowed it to load and link parts of its implementation at runtime.
Freeplane has kept this idea and converted plug-ins from a proprietary format to OSGi. However, during start-up, they are searched in certain directories and use their own class loaders coming from the OSGi Framework. The use of OSGi Framework brings infrastructure for automatic updates and for distributing the plug-ins with all their dependencies not visible to the rest of the code so that they might use their own dependency versions as their dependencies are not visible to any other plug-ins. This complexity explains why Freeplane source code contains a bunch of debugging configurations and why its eclipse project configuration is non-trivial and should be generated using Gradle instead of just importing existing Gradle build scripts.

It could be simplified, however, changes need to be prioritized and designed.

**Reply by HaaiHenkie** [reply]
Hi Dimitry,

Thank you for your answer.

Your following remark surprises me

> ... and why its eclipse project configuration is non-trivial and should be generated using Gradle instead of just importing existing Gradle build scripts.

I have been importing Freeplane source code as a Gradle project into Eclipse for the last couple of years without running the ``cleanEclipse eclipse`` Gradle tasks. I only need to run the ``clean build`` tasks first to prevent some minor import problems. If I do so, all dependencies are resolved and I am able to run Freeplane from the Eclipse IDE. What problems would you expect?

Also importing freeplane_root and the add-on project as a gradle composite project looks very promising, though I haven't used it extensively yet. This is how the dependencies in the add-on's ``build.gradle`` file look like:

```Gradle
    dependencies {
        implementation 'freeplane_root:freeplane' ,
            'freeplane_root:freeplane_api',
            'freeplane_root:freeplane_debughelper',
            'freeplane_root:freeplane_framework',
            'freeplane_root:freeplane_plugin_bugreport',
            'freeplane_root:freeplane_plugin_formula',
            'freeplane_root:freeplane_plugin_jsyntaxpane',
            'freeplane_root:freeplane_plugin_latex',
            'freeplane_root:freeplane_plugin_markdown',
            'freeplane_root:freeplane_plugin_openmaps',
            'freeplane_root:freeplane_plugin_script',
            'freeplane_root:freeplane_plugin_svg',
            'freeplane_root:JOrtho_0.4_freeplane'
    }
```
I suspect you could similarly resolve these dependencies if they were available in a Maven repository. Based on your explanation one would not be able to run Freeplane from the IDE.

Simplifying the technical setup of Freeplane sounds like major refactoring. I suspect add-on and script developers are a relative small group within the Freeplane users community. I think it only is worthwhile when it has other large benefits for Freeplane, e.g. better maintainability, faster development, extension to mobile platforms, etc.

For now I will continue to work with de composite set-up and see if I run into problems or other disadvantages. I might even be able to develop a gradle task that can automatically add and update the Freeplane stable source.

**Comment by EdoFro** [comment]
Hi,
I tried once again to use the [gradle-plugin](https://github.com/freeplane/gradle-plugin) but I couldn't make it work.

I really don't understand how to use IntelliJ Idea to build Add-ons or to build Freeplane. I allways get errors.

@dpolivaev,
May I ask to open the discussion forum in that project, please?
That way all questions and topics will be plugin / add-on related and they won't dirt the discussion in here.

thanks in advance.

edo

**Reply by dpolivaev** [reply]
Done

**Reply by sgjohnson1981** [reply]
Hope you don't mind, @EdoFro, but I went ahead and added the first discussion.
https://github.com/freeplane/gradle-plugin/discussions/6

**Comment by sifran-github** [comment]
I also would like to have some better IDE for development. Until now i use the inline script editor.
Usually I work with VS Code, but any free solution for Freeplane scripting  can also be fine.

I don't know what are the options and what should work.
Guide me on what to try... i can try and see if i can make something work...

If you want we can do some online meeting and try.

**Comment by OmarMohamedAero** [comment]
@HaaiHenkie @dpolivaev @sifran-github @EdoFro 
Thanks for sharing and helping us, and sharing your ideas and work 

 I have been struggling with this for a week, trying to solve this problem. I could build the add-on based on the steps shown [here](https://docs.freeplane.org/scripting/Add-ons_(Develop).html), but it did not work, and I’m facing this issue. If you can, please help me with this, as I’m burned out.
 
 the add-on i created  [here ](https://github.com/OmarAhmed111081/GoGo.git)

[my issue](https://github.com/freeplane/gradle-plugin/discussions/9#discussion-7248226)


I'm disappointed after long time trying and finally failed

**Comment by OmarMohamedAero** [comment]
@macmarrum 
Can you make a video on how to create add-on for freeplane ?

**Reply by macmarrum** [reply]
I don't do videos. But the steps to create an add-on aren't overly complicated.

1. Install **Developer Tools** add-on → https://docs.freeplane.org/getting-started/Add-ons_(install).html#for-developers
1. Create a Groovy script that can be executed by Freeplane, e.g. 
```groovy
// @ExecutionModes({ON_SELECTED_NODE})
def icons = new HashSet<String>()
c.findAll().each { icons.addAll(it.icons.icons) }
ui.showMessage(icons.join('\n'), 1)
```
2. Save it to `<user-dir>/scripts/reportIcons.groovy`
3. Restart Freeplane and execute the script to make sure it works OK `Tools->Scripts->Report Icons`
4. Create a new mind map
5. Execute `Tools->Developer Tools->Build add-on`
6. Save the mind map to disk. Create a `scripts` directory next to the mind map.
7. Move your Groovy script to the `scripts` directory (from `<user-dir>/scripts/reportIcons.groovy`)
8. The mind map becomes a template for your add-on. Follow the guidelines in notes to populate the mind map, especially in the `scripts` node, where you need to create a new node with the name of your script (here `reportIcons.groovy`)
You can have a look at an existing add-on as an example, e.g. [PackAndGoGo-v2024.3.19.addon.mm](https://github.com/macmarrum/freeplane-PackAndGoGo/releases/download/v2024.3.19/PackAndGoGo-v2024.3.19.addon.mm) -- open it in Freeplane and choose `No` when prompted to install it
9. Once your add-on template is populated, run `Tools->Developer Tools->Package add-on for publication`

**Reply by OmarMohamedAero** [reply]
@macmarrum 

Thanks a lot, @macmarrum . I followed the steps, but the script gave me an error. I hope you have time to help me solve it. I have uploaded all the files and the script [here](https://github.com/OmarAhmed111081/addon1.git).

Thank you for your time and help!


error window 
![image](https://github.com/user-attachments/assets/9941c723-88c8-4392-ae1a-d09082758cee)

**Reply by HaaiHenkie** [reply]
This is an error in you script and has nothing to do with your original question regarding creating an add-on. I get an error when I follow your link, so I cannot inspect the script. If your goal is to learn how to create an add-on, try it with a working script or add-on

**Reply by OmarMohamedAero** [reply]
I'm sorry here is the link for the files 

[link ](https://github.com/OmarAhmed111081/addon1.git)

I was able to create an add-on, and it works in some cases, but sometimes it gives an error.



I'm sorry for bothering both of you.

**Comment by OmarMohamedAero** [comment]
@macmarrum 

I have a question. The scripts worked, but I need to stream data instead of collecting it. When the addon operates, it loads the entire dataset at once, which creates a very large array and causes processing issues.

Can you help me solve this?

[Link for the script](https://github.com/OmarAhmed111081/addon1.git)

**Reply by dpolivaev** [reply]
Hi  @OmarAhmed111081 , I’ve noticed you tag users quite frequently. To keep the conversation respectful and flowing smoothly, please avoid tagging others unless there hasn't been a response after a reasonable amount of time. This helps create a more considerate environment for everyone. Thanks for your understanding!

**Reply by OmarMohamedAero** [reply]
Sorry about that, but this problem is causing a big issue.

I don’t mean to bother anyone, but I’ve really spent a lot of time working on this problem.

**Reply by macmarrum** [reply]
> The scripts worked, but I need to stream data instead of collecting it. When the addon operates, it loads the entire dataset at once, which creates a very large array and causes processing issues.

I think this is a general Java question rather then a Freeplane one. I have limited time and knowledge to be able to help you with that. You might get better support in Stack Overflow or other Java-focused communities.

**Comment by buckeye43210** [comment]
Are there any plans to publish Freeplane to Maven Central? My employer has an automated process to pull artifacts from Maven, PyPi, etc. Having Freeplane in Maven Central would make on-boarding Freeplane much easier for us.
